subsys/net/lib/config/init_clock_sntp.c

```diff
@@ -112,10 +112,230 @@ int net_init_clock_via_sntp(void)
+	}
+}
+
+static int sntp_query_async(struct net_sockaddr *addr, net_socklen_t addrlen)
+{
+	int ret;
+
+	ret = sntp_init_async(&sntp_async_ctx, addr, addrlen, &service_sntp_async);
+	if (ret < 0) {
+		LOG_ERR("Failed to initialize SNTP context (%d)", ret);
+		goto end;
+	}
+
+	ret = sntp_send_async(&sntp_async_ctx);
+	if (ret < 0) {
+		LOG_ERR("Failed to send SNTP query (%d)", ret);
+		sntp_close_async(&service_sntp_async);
+		goto end;
+	}
+
+	k_work_reschedule(&sntp_async_timeout_work,
+			  K_MSEC(CONFIG_NET_CONFIG_SNTP_INIT_TIMEOUT));
+
+end:
+	return ret;
+}
+
+static void dns_result_cb(enum dns_resolve_status status,
+			  struct dns_addrinfo *info,
+			  void *user_data)
+{
+	int ret;
+
+	if (status == DNS_EAI_CANCELED || status == DNS_EAI_FAIL) {
+		/* If IPv4 query failed, try IPv6. Otherwise, just schedule next retry. */
+		if (sntp_addr.ss_family == NET_AF_INET && IS_ENABLED(CONFIG_NET_IPV6)) {
+			sntp_addr.ss_family = NET_AF_INET6;
+			sntp_addrlen = 0;
+
+			ret = dns_get_addr_info(CONFIG_NET_CONFIG_SNTP_INIT_SERVER,
+						DNS_QUERY_TYPE_AAAA, NULL, dns_result_cb,
+						NULL, CONFIG_NET_CONFIG_SNTP_INIT_TIMEOUT);
+			if (ret == 0) {
+				return;
+			}
+		}
+
+		LOG_WRN("DNS query timed out");
+
+		k_work_reschedule(&sntp_resync_work_handle, RESYNC_FAILED_INTERVAL);
+
+		return;
+	}
+
+	if (status == DNS_EAI_ALLDONE) {
+		/* If address was found, schedule SNTP query, if that fails schedule next retry. */
+		ret = sntp_query_async(net_sad(&sntp_addr), sntp_addrlen);
+		if (ret < 0) {
+			k_work_reschedule(&sntp_resync_work_handle, RESYNC_FAILED_INTERVAL);
+		}
+
+		return;
+	}
+
+	if (status == DNS_EAI_INPROGRESS && info != NULL) {
+		switch (info->ai_family) {
+		case NET_AF_INET:
+			if (sntp_addrlen > 0) {
+				/* Already got address, skip others. */
+				break;
+			}
+
+			sntp_addrlen = info->ai_addrlen;
+			sntp_addr.ss_family = NET_AF_INET;
+			net_ipv4_addr_copy_raw(net_sin(net_sad(&sntp_addr))->sin_addr.s4_addr,
+					       net_sin(&info->ai_addr)->sin_addr.s4_addr);
+			net_sin(net_sad(&sntp_addr))->sin_port = net_htons(123);
+			break;
+
+		case NET_AF_INET6:
+			if (sntp_addrlen > 0) {
+				/* Already got address, skip others. */
+				break;
+			}
+
+			sntp_addrlen = info->ai_addrlen;
+			sntp_addr.ss_family = NET_AF_INET6;
+			net_ipv6_addr_copy_raw(net_sin6(net_sad(&sntp_addr))->sin6_addr.s6_addr,
+					       net_sin6(&info->ai_addr)->sin6_addr.s6_addr);
+			net_sin6(net_sad(&sntp_addr))->sin6_port = net_htons(123);
+			break;
+
+		default:
+			/* Ignore other records. */
+			break;
+		}
+	}
+}
+
+static int dns_query_async(void)
+{
+	enum dns_query_type type;
+	int ret;
+
+	memset(&sntp_addr, 0, sizeof(sntp_addr));
+	sntp_addrlen = 0;
+
+	if (IS_ENABLED(CONFIG_NET_IPV4)) {
+		sntp_addr.ss_family = NET_AF_INET;
+		type = DNS_QUERY_TYPE_A;
+	} else {
+		sntp_addr.ss_family = NET_AF_INET6;
+		type = DNS_QUERY_TYPE_AAAA;
+	}
+
+	ret = dns_get_addr_info(CONFIG_NET_CONFIG_SNTP_INIT_SERVER,
```

Comment:
This also works when DNS is enabled, but the Kconfig is a IP address right?
Just like `getaddrinfo()`